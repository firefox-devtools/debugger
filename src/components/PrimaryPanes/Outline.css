.outline {
  overflow-y: auto;
}

.outline .outline-pane-info {
  width: 100%;
  font-style: italic;
  text-align: center;
  padding: 0.5em;
  user-select: none;
  font-size: 12px;
}

.outline-list {
  list-style-type: none;
  padding-left: 0px;
  width: 100%;
}

.outline-list__element {
  color: blue;
  padding-bottom: 0.2rem;
  padding-left: 1rem;
  padding-right: 0.5rem;
  padding-top: 0.2rem;
  cursor: default;
}

.outline-list__element:hover {
  background: var(--theme-toolbar-background-hover);
}
