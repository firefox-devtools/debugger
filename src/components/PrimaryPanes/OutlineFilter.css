.outline-filter {
  margin: 5px 0 0 0;
  padding: 0px 10px;
}

.outline-filter-input {
  padding: 0.5em 1.6em;
  width: 100%;
  border: 1px solid var(--theme-splitter-color);
  background-color: var(--theme-sidebar-background);
  color: var(--theme-body-color);
  font-size: 12px;
  -moz-user-select: text;
  user-select: text;
}

.outline-filter-input.focused {
  border: 1px solid var(--theme-highlight-blue);
}

.outline-filter-input::placeholder {
  font-style: italic;
  color: var(--theme-comment);
}
