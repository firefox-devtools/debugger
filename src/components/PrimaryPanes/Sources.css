.sources-panel {
  background-color: var(--theme-sidebar-background);
  display: flex;
  flex: 1;
  flex-direction: column;
  overflow: hidden;
  position: relative;
}

.sources-panel * {
  -moz-user-select: none;
  user-select: none;
}

.sources-header {
  height: 29px;
  background-color: var(--theme-toolbar-background);
  border-bottom: 1px solid var(--theme-splitter-color);
  padding-top: 0px;
  padding-bottom: 0px;
  line-height: 30px;
  font-size: 1.2em;
  display: flex;
  align-items: baseline;
  -moz-user-select: none;
  user-select: none;
  justify-content: flex-end;
}

.theme-dark .sources-header {
  background-color: var(--theme-tab-toolbar-background);
}

.sources-header {
  padding-inline-start: 10px;
}

.sources-header-info {
  font-size: 12px;
  color: var(--theme-comment-alt);
  font-weight: lighter;
  white-space: nowrap;
  padding-inline-end: 10px;
}

.sources-list {
  flex: 1;
  display: flex;
  overflow-x: hidden;
  overflow-y: auto;
}

.sources-list .managed-tree {
  flex: 1;
  display: flex;
  overflow-x: hidden;
  overflow-y: auto;
}

.theme-dark .sources-list .tree .node:not(.focused) svg {
  fill: var(--theme-content-color3);
}

.theme-dark .source-list .tree .node.focused {
  background-color: var(--theme-tab-toolbar-background);
}

.no-sources-message {
  font-size: 12px;
  color: var(--theme-comment-alt);
  font-weight: lighter;
  padding-top: 5px;
  flex-grow: 1;
  display: flex;
  justify-content: center;
  align-items: center;
}

.sources-panel .source-footer {
  position: relative;
}

.sources-panel .outline {
  display: flex;
  flex: 1;
}

.sources-panel .outline.hidden {
  display: none;
}

.hidden {
  display: none;
}

.source-footer {
  width: 100%;
}

.source-footer .tab {
  flex: 1;
  justify-content: center;
  border: 1px solid transparent;
  border-bottom-left-radius: 2px;
  border-bottom-right-radius: 2px;
  display: inline-flex;
  position: relative;
  transition: all 0.25s ease;
  overflow: hidden;
  padding: 5px;
  margin-bottom: 0px;
  margin-top: -1px;
}

.source-footer .tab:hover {
  background-color: var(--theme-toolbar-background-alt);
  border-color: var(--theme-splitter-color);
}

.source-footer .tab.active {
  color: var(--theme-body-color);
  background-color: var(--theme-toolbar-background);
  border-color: var(--theme-splitter-color);
}

.source-footer .tab.active path,
.source-footer .tab:hover path {
  fill: var(--theme-body-color);
}
