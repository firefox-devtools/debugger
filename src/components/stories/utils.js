const renderComponentFromFixture =
      require("../../test/utils/renderComponentFromFixture");

const { storiesOf } = require("@kadira/storybook");

module.exports = {
  renderComponent: renderComponentFromFixture,
  storiesOf
};
