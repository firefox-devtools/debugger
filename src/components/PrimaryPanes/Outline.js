// @flow

import React, { Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import classnames from "classnames";
import actions from "../../actions";
import { getSelectedSource, getSymbols } from "../../selectors";
import { isEnabled } from "devtools-config";
import "./Outline.css";
import previewFunction from "../shared/previewFunction";

import type {
  SymbolDeclarations,
  SymbolDeclaration
} from "../../utils/parser/getSymbols";
import type { SourceRecord } from "../../reducers/sources";

type State = any;
type Props = {
  isHidden: boolean,
  symbols: SymbolDeclarations,
  selectSource: (string, { line: number }) => void,
  selectedSource: ?SourceRecord
};

class Outline extends Component<Props, State> {
  constructor(props) {
    super(props);
    this.state = {};
  }

  selectItem(location) {
    const { selectedSource, selectSource } = this.props;
    if (!selectedSource) {
      return;
    }
    const selectedSourceId = selectedSource.get("id");
    const startLine = location.start.line;
    selectSource(selectedSourceId, { line: startLine });
  }

  renderFunction(func: SymbolDeclaration) {
    const { name, location } = func;

    return (
      <li
        key={`${name}:${location.start.line}:${location.start.column}`}
        className="outline-list__element"
        onClick={() => this.selectItem(location)}
      >
        {previewFunction({ name })}
      </li>
    );
  }

  renderFunctions() {
    const { symbols } = this.props;

    return symbols.functions
      .filter(func => func.name != "anonymous")
      .map(func => this.renderFunction(func));
  }

  render() {
    const { isHidden } = this.props;
    if (!isEnabled("outline")) {
      return null;
    }

    return (
      <div className={classnames("outline", { hidden: isHidden })}>
        <ul className="outline-list">
          {this.renderFunctions()}
        </ul>
      </div>
    );
  }
}

Outline.displayName = "Outline";

export default connect(
  state => {
    const selectedSource = getSelectedSource(state);
    return {
      symbols: getSymbols(state, selectedSource && selectedSource.toJS()),
      selectedSource
    };
  },
  dispatch => bindActionCreators(actions, dispatch)
)(Outline);
