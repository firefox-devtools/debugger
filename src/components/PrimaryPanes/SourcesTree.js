/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

// Dependencies
import React, { Component } from "react";
import classnames from "classnames";
import { showMenu } from "devtools-contextmenu";
import { connect } from "react-redux";

// Selectors
import {
  getShownSource,
  getSelectedSource,
  getDebuggeeUrl,
  getExpandedState,
  getProjectDirectoryRoot,
  getSources
} from "../../selectors";

// Actions
import { setExpandedState } from "../../actions/source-tree";
import { selectLocation } from "../../actions/sources";
import {
  setProjectDirectoryRoot,
  clearProjectDirectoryRoot
} from "../../actions/ui";

// Components
import ManagedTree from "../shared/ManagedTree";
import Svg from "../shared/Svg";

// Utils
import {
  createTree,
  getDirectories,
  isDirectory,
  nodeHasChildren,
  updateTree
} from "../../utils/sources-tree";

import { getRawSourceURL, getSourceClassnames } from "../../utils/source";
import { copyToTheClipboard } from "../../utils/clipboard";
import { features } from "../../utils/prefs";

import type { SourcesMap } from "../../reducers/types";
import type { SourceRecord } from "../../types";

type Props = {
  selectLocation: Object => void,
  setExpandedState: any => void,
  setProjectDirectoryRoot: string => void,
  clearProjectDirectoryRoot: void => void,
  sources: SourcesMap,
  shownSource?: string,
  selectedSource?: SourceRecord,
  debuggeeUrl: string,
  projectRoot: string,
  expanded?: any
};

type State = {
  focusedItem?: any,
  parentMap: any,
  sourceTree: Object,
  projectRoot: string,
  uncollapsedTree: any,
  listItems?: any,
  highlightItems?: any
};

class SourcesTree extends Component<Props, State> {
  focusItem: Function;
  selectItem: Function;
  getPath: Function;
  getIcon: Function;
  queueUpdate: Function;
  onContextMenu: Function;
  renderItem: Function;
  mounted: boolean;

  constructor(props) {
    super(props);
    const { debuggeeUrl, sources, projectRoot } = this.props;

    this.state = createTree({
      projectRoot,
      debuggeeUrl,
      sources
    });
  }

  componentWillReceiveProps(nextProps) {
    const {
      projectRoot,
      debuggeeUrl,
      sources,
      shownSource,
      selectedSource
    } = this.props;

    const { uncollapsedTree, sourceTree } = this.state;

    if (
      projectRoot != nextProps.projectRoot ||
      debuggeeUrl != nextProps.debuggeeUrl ||
      nextProps.sources.size === 0
    ) {
      // early recreate tree because of changes
      // to project root, debugee url or lack of sources
      return this.setState(
        createTree({
          sources: nextProps.sources,
          debuggeeUrl: nextProps.debuggeeUrl,
          projectRoot: nextProps.projectRoot
        })
      );
    }
    if (nextProps.shownSource && nextProps.shownSource != shownSource) {
      const listItems = getDirectories(nextProps.shownSource, sourceTree);

      if (listItems && listItems[0]) {
        this.selectItem(listItems[0]);
      }

      return this.setState({ listItems });
    }

    if (
      nextProps.selectedSource &&
      nextProps.selectedSource != selectedSource
    ) {
      const highlightItems = getDirectories(
        getRawSourceURL(nextProps.selectedSource.get("url")),
        sourceTree
      );

      return this.setState({ highlightItems });
    }

    // NOTE: do not run this every time a source is clicked,
    // only when a new source is added
    if (nextProps.sources != this.props.sources) {
      this.setState(
        updateTree({
          newSources: nextProps.sources,
          prevSources: sources,
          debuggeeUrl,
          projectRoot,
          uncollapsedTree,
          sourceTree
        })
      );
    }
  }

  focusItem = item => {
    this.setState({ focusedItem: item });
  };

  selectItem = item => {
    if (!nodeHasChildren(item)) {
      this.props.selectLocation({ sourceId: item.contents.get("id") });
    }
  };

  getPath = item => {
    const { sources } = this.props;
    const obj = item.contents.get && item.contents.get("id");

    let blackBoxedPart = "";

    if (
      typeof obj !== "undefined" &&
      sources.has(obj) &&
      sources.get(obj).get("isBlackBoxed")
    ) {
      blackBoxedPart = "update";
    }

    return `${item.path}/${item.name}/${blackBoxedPart}`;
  };

  getIcon = (sources, item, depth) => {
    const { debuggeeUrl, projectRoot } = this.props;

    if (item.path === "webpack://") {
      return <Svg name="webpack" />;
    }
    if (item.path === "/Angular") {
      return <Svg name="angular" />;
    }

    if (depth === 0 && projectRoot === "") {
      return (
        <img
          className={classnames("domain", {
            debuggee: debuggeeUrl && debuggeeUrl.includes(item.name)
          })}
        />
      );
    }

    if (!nodeHasChildren(item)) {
      const obj = item.contents.get("id");
      const source = sources.get(obj);
      return (
        <img
          className={classnames(
            getSourceClassnames(source.toJS()),
            "source-icon"
          )}
        />
      );
    }

    return <img className="folder" />;
  };

  onContextMenu = (event, item) => {
    const copySourceUri2Label = L10N.getStr("copySourceUri2");
    const copySourceUri2Key = L10N.getStr("copySourceUri2.accesskey");
    const setDirectoryRootLabel = L10N.getStr("setDirectoryRoot.label");
    const setDirectoryRootKey = L10N.getStr("setDirectoryRoot.accesskey");
    const removeDirectoryRootLabel = L10N.getStr("removeDirectoryRoot.label");

    event.stopPropagation();
    event.preventDefault();

    const menuOptions = [];

    if (!isDirectory(item)) {
      const source = item.contents.get("url");
      const copySourceUri2 = {
        id: "node-menu-copy-source",
        label: copySourceUri2Label,
        accesskey: copySourceUri2Key,
        disabled: false,
        click: () => copyToTheClipboard(source)
      };

      menuOptions.push(copySourceUri2);
    }

    if (isDirectory(item) && features.root) {
      const { path } = item;
      const { projectRoot } = this.props;

      if (projectRoot.endsWith(path)) {
        menuOptions.push({
          id: "node-remove-directory-root",
          label: removeDirectoryRootLabel,
          disabled: false,
          click: () => this.props.clearProjectDirectoryRoot()
        });
      } else {
        menuOptions.push({
          id: "node-set-directory-root",
          label: setDirectoryRootLabel,
          accesskey: setDirectoryRootKey,
          disabled: false,
          click: () => this.props.setProjectDirectoryRoot(path)
        });
      }
    }

    showMenu(event, menuOptions);
  };

  renderItem = (item, depth, focused, _, expanded, { setExpanded }) => {
    const arrow = nodeHasChildren(item) ? (
      <img
        className={classnames("arrow", {
          expanded: expanded
        })}
      />
    ) : (
      <i className="no-arrow" />
    );

    const { sources } = this.props;
    const icon = this.getIcon(sources, item, depth);

    return (
      <div
        className={classnames("node", { focused })}
        key={item.path}
        onClick={e => {
          this.focusItem(item);

          if (isDirectory(item)) {
            setExpanded(item, !!expanded, e.altKey);
          } else {
            this.selectItem(item);
          }
        }}
        onContextMenu={e => this.onContextMenu(e, item)}
      >
        {arrow}
        {icon}
        <span className="label"> {this.renderItemName(item.name)} </span>
      </div>
    );
  };

  renderItemName(name) {
    const hosts = {
      "ng://": "Angular",
      "webpack://": "Webpack"
    };

    return hosts[name] || name;
  }

  renderEmptyElement(message) {
    return <div className="no-sources-message">{message}</div>;
  }

  render() {
    const { expanded, projectRoot } = this.props;
    const {
      focusedItem,
      highlightItems,
      listItems,
      parentMap,
      sourceTree
    } = this.state;

    const onExpand = (item, expandedState) => {
      this.props.setExpandedState(expandedState);
    };

    const onCollapse = (item, expandedState) => {
      this.props.setExpandedState(expandedState);
    };

    const isEmpty = sourceTree.contents.length === 0;

    const isCustomRoot = projectRoot !== "";
    let roots = () => sourceTree.contents;

    let clearProjectRootButton = null;

    // The "sourceTree.contents[0]" check ensures that there are contents
    // A custom root with no existing sources will be ignored
    if (isCustomRoot) {
      let rootLabel = projectRoot.split("/").pop();
      if (sourceTree.contents[0]) {
        rootLabel = sourceTree.contents[0].name;
        roots = () => sourceTree.contents[0].contents;
      }

      clearProjectRootButton = (
        <button
          className="sources-clear-root"
          onClick={() => this.props.clearProjectDirectoryRoot()}
          title={L10N.getStr("removeDirectoryRoot.label")}
        >
          <Svg name="home" />
          <Svg name="breadcrumb" class />
          <span className="sources-clear-root-label">{rootLabel}</span>
        </button>
      );
    }

    if (isEmpty && !isCustomRoot) {
      return this.renderEmptyElement(L10N.getStr("sources.noSourcesAvailable"));
    }

    const treeProps = {
      autoExpandAll: false,
      autoExpandDepth: expanded ? 0 : 1,
      autoExpandOnHighlight: true,
      expanded,
      getChildren: item => (nodeHasChildren(item) ? item.contents : []),
      getParent: item => parentMap.get(item),
      getPath: this.getPath,
      getRoots: roots,
      highlightItems,
      itemHeight: 21,
      key: isEmpty ? "empty" : "full",
      listItems,
      onCollapse,
      onExpand,
      onFocus: this.focusItem,
      renderItem: this.renderItem
    };

    const tree = <ManagedTree {...treeProps} />;

    const onKeyDown = e => {
      if (e.keyCode === 13 && focusedItem) {
        this.selectItem(focusedItem);
      }
    };

    return (
      <div
        className={classnames("sources-pane", {
          "sources-list-custom-root": isCustomRoot
        })}
      >
        {isCustomRoot ? (
          <div className="sources-clear-root-container">
            {clearProjectRootButton}
          </div>
        ) : null}
        {isEmpty ? (
          this.renderEmptyElement(L10N.getStr("sources.noSourcesAvailableRoot"))
        ) : (
          <div className="sources-list" onKeyDown={onKeyDown}>
            {tree}
          </div>
        )}
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    shownSource: getShownSource(state),
    selectedSource: getSelectedSource(state),
    debuggeeUrl: getDebuggeeUrl(state),
    expanded: getExpandedState(state),
    projectRoot: getProjectDirectoryRoot(state),
    sources: getSources(state)
  };
};

const actionCreators = {
  setExpandedState,
  selectLocation,
  setProjectDirectoryRoot,
  clearProjectDirectoryRoot
};

export default connect(mapStateToProps, actionCreators)(SourcesTree);
