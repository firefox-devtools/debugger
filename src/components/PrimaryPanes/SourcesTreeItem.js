/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import React, { Component } from "react";
import { connect } from "react-redux";
import classnames from "classnames";
import { showMenu } from "devtools-contextmenu";

import SourceIcon from "../shared/SourceIcon";
import Svg from "../shared/Svg";

import { getGeneratedSourceByURL } from "../../selectors";
import actions from "../../actions";

import { isOriginal as isOriginalSource } from "../../utils/source";
import { isDirectory } from "../../utils/sources-tree";
import { copyToTheClipboard } from "../../utils/clipboard";
import { features } from "../../utils/prefs";

import type { TreeNode } from "../../utils/sources-tree/types";
import type { Source } from "../../types";

type Props = {
  debuggeeUrl: string,
  projectRoot: string,
  source: ?Source,
  item: TreeNode,
  depth: number,
  focused: boolean,
  expanded: boolean,
  hasMatchingGeneratedSource: boolean,
  setExpanded: (TreeNode, boolean, boolean) => void,
  focusItem: TreeNode => void,
  selectItem: TreeNode => void,
  clearProjectDirectoryRoot: () => void,
  setProjectDirectoryRoot: string => void
};

type State = {};

class SourceTreeItem extends Component<Props, State> {
  getIcon(item: TreeNode, depth: number) {
    const { debuggeeUrl, projectRoot, source } = this.props;

    if (item.path === "webpack://") {
      return <Svg name="webpack" />;
    } else if (item.path === "ng://") {
      return <Svg name="angular" />;
    } else if (item.path.startsWith("moz-extension://") && depth === 0) {
      return <img className="extension" />;
    }

    if (depth === 0 && projectRoot === "") {
      return (
        <img
          className={classnames("domain", {
            debuggee: debuggeeUrl && debuggeeUrl.includes(item.name)
          })}
        />
      );
    }

    if (isDirectory(item)) {
      return <img className="folder" />;
    }

    if (source) {
      return <SourceIcon source={source} />;
    }

    return null;
  }

  onClick = (e: MouseEvent) => {
    const { expanded, item, focusItem, setExpanded, selectItem } = this.props;

    focusItem(item);

    if (isDirectory(item)) {
      setExpanded(item, !!expanded, e.altKey);
    } else {
      selectItem(item);
    }
  };

  onContextMenu = (event: Event, item: TreeNode) => {
    const copySourceUri2Label = L10N.getStr("copySourceUri2");
    const copySourceUri2Key = L10N.getStr("copySourceUri2.accesskey");
    const setDirectoryRootLabel = L10N.getStr("setDirectoryRoot.label");
    const setDirectoryRootKey = L10N.getStr("setDirectoryRoot.accesskey");
    const removeDirectoryRootLabel = L10N.getStr("removeDirectoryRoot.label");

    event.stopPropagation();
    event.preventDefault();

    const menuOptions = [];

    if (!isDirectory(item)) {
      // Flow requires some extra handling to ensure the value of contents.
      const { contents } = item;

      if (!Array.isArray(contents)) {
        const copySourceUri2 = {
          id: "node-menu-copy-source",
          label: copySourceUri2Label,
          accesskey: copySourceUri2Key,
          disabled: false,
          click: () => copyToTheClipboard(contents.url)
        };

        menuOptions.push(copySourceUri2);
      }
    }

    if (isDirectory(item) && features.root) {
      const { path } = item;
      const { projectRoot } = this.props;

      if (projectRoot.endsWith(path)) {
        menuOptions.push({
          id: "node-remove-directory-root",
          label: removeDirectoryRootLabel,
          disabled: false,
          click: () => this.props.clearProjectDirectoryRoot()
        });
      } else {
        menuOptions.push({
          id: "node-set-directory-root",
          label: setDirectoryRootLabel,
          accesskey: setDirectoryRootKey,
          disabled: false,
          click: () => this.props.setProjectDirectoryRoot(path)
        });
      }
    }

    showMenu(event, menuOptions);
  };

  renderItemArrow() {
    const { item, expanded } = this.props;
    return isDirectory(item) ? (
      <img className={classnames("arrow", { expanded })} />
    ) : (
      <i className="no-arrow" />
    );
  }

  renderItemName() {
    const { item } = this.props;

    switch (item.name) {
      case "ng://":
        return "Angular";
      case "webpack://":
        return "Webpack";
      default:
        return `${item.name}`;
    }
  }

  render() {
    const { item, depth, focused, hasMatchingGeneratedSource } = this.props;
    const suffix = hasMatchingGeneratedSource ? (
      <span className="suffix">{L10N.getStr("sourceFooter.mappedSuffix")}</span>
    ) : null;

    return (
      <div
        className={classnames("node", { focused })}
        key={item.path}
        onClick={this.onClick}
        onContextMenu={e => this.onContextMenu(e, item)}
      >
        {this.renderItemArrow()}
        {this.getIcon(item, depth)}
        <span className="label">
          {" "}
          {this.renderItemName()} {suffix}
        </span>
      </div>
    );
  }
}

function getHasMatchingGeneratedSource(state, source: ?Source) {
  if (!source || !isOriginalSource(source)) {
    return false;
  }

  return !!getGeneratedSourceByURL(state, source.url);
}

const mapStateToProps = (state, props) => {
  const { source } = props;
  return {
    hasMatchingGeneratedSource: getHasMatchingGeneratedSource(state, source)
  };
};

export default connect(
  mapStateToProps,
  {
    setProjectDirectoryRoot: actions.setProjectDirectoryRoot,
    clearProjectDirectoryRoot: actions.clearProjectDirectoryRoot
  }
)(SourceTreeItem);
