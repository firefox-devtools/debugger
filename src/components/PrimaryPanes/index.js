// @flow

import { DOM as dom, Component, createFactory } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { formatKeyShortcut } from "../../utils/text";
import actions from "../../actions";
import { getSources, getActiveSearchState } from "../../selectors";
import { isEnabled } from "devtools-config";
import "./Sources.css";
import classnames from "classnames";

import _Outline from "./Outline";
const Outline = createFactory(_Outline);

import _SourcesTree from "./SourcesTree";
const SourcesTree = createFactory(_SourcesTree);

import type { SourcesMap } from "../../reducers/types";
type SourcesState = {
  selectedPane: string
};

type State = SourcesState;
type Props = {
  sources: SourcesMap,
  selectSource: (string, Object) => void,
  horizontal: boolean,
  setActiveSearch: string => void,
  closeActiveSearch: () => void,
  sourceSearchOn: boolean
};

class PrimaryPanes extends Component<Props, State> {
  renderShortcut: Function;
  selectedPane: String;
  showPane: Function;
  renderFooter: Function;
  renderChildren: Function;

  constructor(props: Props) {
    super(props);
    this.state = { selectedPane: "sources" };

    this.renderShortcut = this.renderShortcut.bind(this);
    this.showPane = this.showPane.bind(this);
    this.renderFooter = this.renderFooter.bind(this);
  }

  showPane(selectedPane: string) {
    this.setState({ selectedPane });
  }

  renderOutlineTabs() {
    if (!isEnabled("outline")) {
      return;
    }

    return [
      dom.div(
        {
          className: classnames("tab", {
            active: this.state.selectedPane === "sources"
          }),
          onClick: () => this.showPane("sources"),
          key: "sources-tab"
        },
        "Sources View"
      ),
      dom.div(
        {
          className: classnames("tab", {
            active: this.state.selectedPane === "outline"
          }),
          onClick: () => this.showPane("outline"),
          key: "outline-tab"
        },
        "Outline View"
      )
    ];
  }

  renderFooter() {
    return dom.div(
      {
        className: "source-footer"
      },
      this.renderOutlineTabs()
    );
  }

  renderShortcut() {
    if (this.props.horizontal) {
      return dom.span(
        {
          className: "sources-header-info",
          dir: "ltr",
          onClick: () => {
            if (this.props.sourceSearchOn) {
              return this.props.closeActiveSearch();
            }
            this.props.setActiveSearch("source");
          }
        },
        L10N.getFormatStr(
          "sources.search",
          formatKeyShortcut(L10N.getStr("sources.search.key2"))
        )
      );
    }
  }

  renderHeader() {
    return dom.div({ className: "sources-header" }, this.renderShortcut());
  }

  render() {
    const { selectedPane } = this.state;
    const { sources, selectSource } = this.props;

    return dom.div(
      { className: "sources-panel" },
      this.renderHeader(),
      SourcesTree({
        sources,
        selectSource,
        isHidden: selectedPane === "outline"
      }),
      Outline({ selectSource, isHidden: selectedPane === "sources" }),
      this.renderFooter()
    );
  }
}

PrimaryPanes.displayName = "PrimaryPanes";

export default connect(
  state => ({
    sources: getSources(state),
    sourceSearchOn: getActiveSearchState(state) === "source"
  }),
  dispatch => bindActionCreators(actions, dispatch)
)(PrimaryPanes);
