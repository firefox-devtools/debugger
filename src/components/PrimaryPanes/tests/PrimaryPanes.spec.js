/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import React from "react";
import { shallow } from "enzyme";
import { showMenu } from "devtools-contextmenu";

import { copyToTheClipboard } from "../../../utils/clipboard";
import PrimaryPanes from "..";

jest.mock("devtools-contextmenu", () => ({ showMenu: jest.fn() }));
jest.mock("../../../utils/clipboard", () => ({
  copyToTheClipboard: jest.fn()
}));

describe("PrimaryPanes", () => {
  afterEach(() => {
    (copyToTheClipboard: any).mockClear();
    showMenu.mockClear();
  });

  describe("with custom root", () => {
    it("renders custom root source list", async () => {
      const { component } = render({
        projectRoot: "mdn.com"
      });
      expect(component).toMatchSnapshot();
    });

    it("calls clearProjectDirectoryRoot on click", async () => {
      const { component, props } = render({
        projectRoot: "mdn"
      });
      component.find(".sources-clear-root").simulate("click");
      expect(props.clearProjectDirectoryRoot).toHaveBeenCalled();
    });

    it("renders empty custom root source list", async () => {
      const { component } = render({
        projectRoot: "custom",
        sources: {}
      });
      expect(component).toMatchSnapshot();
    });
  });
});

function generateDefaults(overrides) {
  return {
    horizontal: false,
    projectRoot: "",
    sourceSearchOn: false,
    setPrimaryPaneTab: jest.fn(),
    setActiveSearch: jest.fn(),
    closeActiveSearch: jest.fn(),
    clearProjectDirectoryRoot: jest.fn(),
    threads: [],
    ...overrides
  };
}

function render(overrides = {}) {
  const props = generateDefaults(overrides);
  // $FlowIgnore
  const component = shallow(<PrimaryPanes.WrappedComponent {...props} />);
  const defaultState = component.state();
  const instance = component.instance();

  instance.shouldComponentUpdate = () => true;

  return { component, props, defaultState, instance };
}
