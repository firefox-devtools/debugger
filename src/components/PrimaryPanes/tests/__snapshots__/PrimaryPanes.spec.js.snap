// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryPanes with custom root renders custom root source list 1`] = `
<Tabs
  activeIndex={1}
  className="sources-panel"
  onActivateTab={[Function]}
>
  <TabList
    activeIndex={0}
    className="source-outline-tabs"
    onActivateTab={[Function]}
    vertical={false}
  >
    <Tab
      active={false}
      className="tab sources-tab"
      key="sources-tab"
    >
      Sources
    </Tab>
    <Tab
      active={false}
      className="tab outline-tab"
      key="outline-tab"
    >
      Outline
    </Tab>
  </TabList>
  <TabPanels
    activeIndex={0}
    className="source-outline-panel has-root"
    hasFocusableContent={true}
  >
    <div>
      <div
        className="sources-clear-root-container"
        key="root"
      >
        <button
          className="sources-clear-root"
          onClick={[Function]}
          title="Remove directory root"
        >
          <AccessibleImage
            className="home"
          />
          <AccessibleImage
            className="breadcrumb"
          />
          <span
            className="sources-clear-root-label"
          >
            mdn.com
          </span>
        </button>
      </div>
    </div>
    <Connect(Outline)
      alphabetizeOutline={false}
      onAlphabetizeClick={[Function]}
    />
  </TabPanels>
</Tabs>
`;

exports[`PrimaryPanes with custom root renders empty custom root source list 1`] = `
<Tabs
  activeIndex={1}
  className="sources-panel"
  onActivateTab={[Function]}
>
  <TabList
    activeIndex={0}
    className="source-outline-tabs"
    onActivateTab={[Function]}
    vertical={false}
  >
    <Tab
      active={false}
      className="tab sources-tab"
      key="sources-tab"
    >
      Sources
    </Tab>
    <Tab
      active={false}
      className="tab outline-tab"
      key="outline-tab"
    >
      Outline
    </Tab>
  </TabList>
  <TabPanels
    activeIndex={0}
    className="source-outline-panel has-root"
    hasFocusableContent={true}
  >
    <div>
      <div
        className="sources-clear-root-container"
        key="root"
      >
        <button
          className="sources-clear-root"
          onClick={[Function]}
          title="Remove directory root"
        >
          <AccessibleImage
            className="home"
          />
          <AccessibleImage
            className="breadcrumb"
          />
          <span
            className="sources-clear-root-label"
          >
            custom
          </span>
        </button>
      </div>
    </div>
    <Connect(Outline)
      alphabetizeOutline={false}
      onAlphabetizeClick={[Function]}
    />
  </TabPanels>
</Tabs>
`;
